src/jacobi/_jacobi.py

```diff
@@ -1 +1,261 @@
+import numpy as np
+import typing as _tp
+from ._typing import Indexable as _Indexable
 
+
+def jacobi(
+    fn: _tp.Callable,
+    x: _tp.Union[float, _Indexable[float]],
+    *args,
+    diagonal: bool = False,
+    method: int = None,
+    mask: np.ndarray = None,
+    rtol: float = 0,
+    maxiter: int = 10,
+    maxgrad: int = 3,
+    step: _tp.Tuple[float, float] = None,
+    diagnostic: dict = None,
+) -> _tp.Tuple[np.ndarray, np.ndarray]:
+    """
+    Return first derivative and its error estimate.
+
+    Parameters
+    ----------
+    fn : Callable
+        Function with the signature `fn(x, *args)`, where `x` is a number or an
+        array of numbers and `*args` are optional auxiliary arguments.
+    x : Number or array of numbers
+        The derivative is computed with respect to `x`. If `x` is an array, the Jacobi
+        matrix is computed with respect to each element of `x`.
+    *args : tuple
+        Additional arguments passed to the function.
+    diagonal : boolean, optional
+        If it is known that the Jacobian computed from the function contains has
+        off-diagonal entries that are all zero, the calculation can be speed up
+        significantly. Set this to true to only compute the diagonal entries of the
+        Jacobi matrix, which are returned as a 1D array. This is faster and uses much
+        less memory if the vector x is very large. Default is False.
+    method : {-1, 0, 1} or None, optional
+        Whether to compute central (0), forward (1) or backward derivatives (-1).
+        The default (None) uses auto-detection.
+    mask : array or None, optional
+        If `x` is an array and `mask` is not None, compute the Jacobi matrix only for the
+        part of the array selected by the mask.
+    rtol : float, optional
+        Relative tolerance for the derivative. The algorithm stops when this relative
+        tolerance is reached. If 0 (the default), the algorithm iterates until the
+        error estimate of the derivative does not improve further.
+    maxiter : int, optional
+        Maximum number of iterations of the algorithm.
+    maxgrad : int, optional
+        Maximum grad of the extrapolation polynomial.
+    step : tuple of float or None, optional
+        Factors that reduce the step size in each iteration relative to the previous
+        step.
+    diagnostic : dict or None, optional
+        If an empty dict is passed to this keyword, it is filled with diagnostic
+        information produced by the algorithm. This reduces performance and is only
+        intended for debugging.
+
+    Returns
+    -------
+    array, array
+        Derivative and its error estimate.
+    """
+    if diagonal:
+        # TODO maybe solve this without introducing a wrapper function
```

Comment:
Yes, this is one of the caveats of diagonal=True. This could also be addressed with the block algorithm I mentioned above, were each block could use a different `method`. In the automatic mode, it is efficient to compute central derivatives by default and only switch to forward/backward derivatives where central derivatives fail.

That being said, for a range of x-values, we could decide to compute the forward derivative for the smallest x value and the backward derivative for the largest x value in automatic mode. That would cover many use cases where the end-points are special.

Reply:
I don't like using a wrapper function, but in principle it is a good design to use some form of recursion to implement a block-wise algorithm.

Reply:
Let's continue this part of the discussion in https://github.com/HDembinski/jacobi/issues/37.